### PR TITLE
8296024: Usage of DirectBuffer::address should be guarded

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes the introduction of **guarding** of the use of `DirectBuffer::address` within the JDK. With the introduction of the Foreign Function and Memory access, it is possible to derive Buffer instances that are backed by native memory that, in turn, can be closed asynchronously by the user (and not only via a `Cleaner` when there is no other reference to the `Buffer` object). If another thread is invoking `MemorySession::close` while a thread is doing work using raw addresses, the outcome is undefined. This means the JVM might crash or even worse, silent modification of unrelated memory might occur. 

Design choices in this PR:

There is already a method `MemorySession::whileAlive` that takes a runnable and that will perform the provided action while acquiring the underlying` MemorySession` (if any). However, using this method entailed relatively large changes while converting larger/nested code segments into lambdas. This would also risk introducing lambda capturing. So, instead, a ~~try-with-resources~~ *try-finally* friendly access method was added. This made is more easy to add guarding and did not risk lambda capturing. Also, introducing lambdas in certain fundamental JDK classes might incur bootstrap problems.

The aforementioned ~~TwR~~ TF is using a ~~"session acquisition" that is not used explicitly in the try block itself~~ session used in the *finally* block. ~~This raises a warning that is suppressed using `@SuppressWarnings("try")`. In the future, we might be able to remove these suppressions by using the reserved variable name `_`.~~

In some cases, where is is guaranteed that the backing memory session is non-closeable, we do not have to guard the use of `DirectBuffer::address`. ~~These cases have been documented in the code.~~

On some occasions, a plurality of acquisitions are made. This would never lead to deadlocks as acquisitions are fundamentally concurrent counters and not resources that only one thread can "own".

I have added comments (and not javadocs) before the declaration of the non-public-api `DirectBuffer::address` method, that the use of the returned address needs to be guarded. It can be discussed if this is preferable or not.

This PR spawns several areas of responsibility and so, I expect more than one reviewer before promoting the PR.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296024](https://bugs.openjdk.org/browse/JDK-8296024): Usage of DirectBuffer::address should be guarded

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [619473ee](https://git.openjdk.org/jdk/pull/11260/files/619473eed389c06501d1bf74f770f50a9a51862c)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [619473ee](https://git.openjdk.org/jdk/pull/11260/files/619473eed389c06501d1bf74f770f50a9a51862c)
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to [619473ee](https://git.openjdk.org/jdk/pull/11260/files/619473eed389c06501d1bf74f770f50a9a51862c)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [619473ee](https://git.openjdk.org/jdk/pull/11260/files/619473eed389c06501d1bf74f770f50a9a51862c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11260/head:pull/11260` \
`$ git checkout pull/11260`

Update a local copy of the PR: \
`$ git checkout pull/11260` \
`$ git pull https://git.openjdk.org/jdk pull/11260/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11260`

View PR using the GUI difftool: \
`$ git pr show -t 11260`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11260.diff">https://git.openjdk.org/jdk/pull/11260.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/nio/Buffer.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/Buffer.java:785]: The throws RuntimeException is not needed here. Also would it be possible to reflow the comment to avoid the wildly long line.

File: src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java:161]: You can drop throws RuntimeEXcepton here too.

File: src/java.base/share/classes/sun/nio/ch/DirectBuffer.java
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/DirectBuffer.java:38]: This comment is very messing and needs to be cleaned up.

File: src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java:252]: This comment is will be confusing to anyone reading this code. Is this really needed?

Comment (minborg) [reply src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java:252]: The reason for the comment is to make it clear why `DirectBuffer::address` can be used directly without guarding. This will also reduce the probability of unnecessary guarding being added in the future. However, if the consensus is that these comments just adds confusion, I am happy to remove them.

File: src/java.base/share/classes/java/util/zip/Adler32.java
Comment (minborg) [reply src/java.base/share/classes/java/util/zip/Adler32.java:102]: Eventually, I hope most of them will be named `_`.

File: src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java
Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java:252]: I'd prefer to see this comment removed from all places that are obviously interacting with the direct buffer cache. These usages are try-finally to acquire and return the temporary direct buffer cache back to the cache. Talking about closable sessions here is definitely confusing.

File: src/java.base/share/classes/java/util/zip/Adler32.java
Comment (mcimadamore) [reply src/java.base/share/classes/java/util/zip/Adler32.java]: maybe just `bufferLock` and and just `acquireBuffer` ?

Comment (AlanBateman) [reply src/java.base/share/classes/java/util/zip/Adler32.java]: Would it be possible to re-examine acquireSession returning Runnable and 
acquireSessionAsAutoCloseable returning AutoCloseable. The naming is awkward at most of the use sites and maybe we can do better.

Comment (minborg) [reply src/java.base/share/classes/java/util/zip/Adler32.java]: I think it would be confusing to have overloads with the same name. 

The `aquireSession(Buffer, boolean)` method is only used once and is used in `IOUtil` so we could rename it to `aquireSessionOrNull` and then rename `acquireSessionAsAutoCloseable` to `acquireSession`.

File: src/java.base/share/classes/sun/nio/ch/DirectBuffer.java
Comment (minborg) [reply src/java.base/share/classes/sun/nio/ch/DirectBuffer.java:41]: Are you happy with the proposed renaming now @ExE-Boss ? I think our propositions crossed in time?

Comment (ExE-Boss) [reply src/java.base/share/classes/sun/nio/ch/DirectBuffer.java:41]: This change was missed in commit <https://github.com/openjdk/jdk/pull/11260/commits/c081b4aebb7b52c7161656b4cba9b0a60d36202a>.

File: src/java.base/share/classes/java/util/zip/Adler32.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/Adler32.java:105]: The updated naming is a bit better but there are it still feels that that there are too many things going on at the use sites ("guard", "session", "reachability fence"). Ideally the acquire would return something with an accessor for the direct buffer address but that may not be possible.  I wonder if changing NOP_CLOSE.close to do reachabilityFence(this) would work as expected and improve many of these use sites?

File: src/java.base/share/classes/sun/nio/fs/NativeBuffers.java
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/fs/NativeBuffers.java:92]: This class, and the temporary buffer cache in sun.nio.ch.Util, are intended to be used with try-finally. There isn't any notion of  asynchronously close so maybe it would best to not add this comment to these sources.

Comment (minborg) [reply src/java.base/share/classes/sun/nio/fs/NativeBuffers.java:92]: That is clear to me but I am trying to prevent future redundant guarding. Anyway, I will remove the comments.

File: src/java.base/share/classes/java/util/zip/Adler32.java
Comment (minborg) [reply src/java.base/share/classes/java/util/zip/Adler32.java:105]: This can certainly be done. We could, for example, add a new method to the `JavaNioAccess` interface:

```AddressAcquisition acquireSession2(Buffer buffer); // to be renamed```

This would allow us to go from:

```
            try (var guard = NIO_ACCESS.acquireSession(buffer)) {
                adler = updateByteBuffer(adler, ((DirectBuffer)buffer).address(), pos, rem);
            } finally {
                Reference.reachabilityFence(buffer);
            }
```

to

```
            try (var guard = NIO_ACCESS.acquireSession2(buffer)) {
                adler = updateByteBuffer(adler, guard.address(), pos, rem);
            }
```

Although this looks much simpler and concise, it means "a new object is created for each invocation" (*). This also allows us to remove the `@SupressWarnings("try")` annotations.

Here is how the undercarriage might look like:

```
                @Override
                public AddressAcquisition acquireSession2(Buffer buffer) {
                    var session = buffer.session();
                    if (session == null) {
                        return AddressAcquisition.create(buffer, () -> {});
                    }
                    session.acquire0();
                    return AddressAcquisition.create(buffer, session::release0);
                }

```

and

```
sealed interface AddressAcquisition extends AutoCloseable {

        long address();

        @Override
        void close();

        static AddressAcquisition create(Buffer buffer, Runnable closeActions) {
            return new AddressAcquisitionImpl(buffer, closeActions);
        }

        final class AddressAcquisitionImpl implements AddressAcquisition {

            private final DirectBuffer directBuffer;
            private final Runnable closeAction;

            public AddressAcquisitionImpl(Buffer buffer,
                                          Runnable closeAction) {
                this.directBuffer = (DirectBuffer) buffer;
                this.closeAction = closeAction;
            }

            @Override
            public long address() {
                return directBuffer.address();
            }

            @Override
            public void close() {
                try {
                    closeAction.run();
                } finally {
                    Reference.reachabilityFence(directBuffer);
                }
            }
        }

    }
```


(*) In reality,  a representation of the object might be allocated on the stack instead.

Comment (AlanBateman) [reply src/java.base/share/classes/java/util/zip/Adler32.java:105]: > Although this looks much simpler and concise, it means "a new object is created for each invocation" 

My comment was actually to see if DirectBuffer could extend AutoCloseable so that the acquire returns "this" for the non-session case.  Doing the equivalent for the session case might leak into MemorySessionImpl implementing DirectBuffer which you probably don't want to do. If NOP_CLOSE.close can do the Reference.reachabilityFence(this) then it would at least improve some of the use-sites.

File: src/java.base/share/classes/java/util/zip/Adler32.java
Comment (mcimadamore) [reply src/java.base/share/classes/java/util/zip/Adler32.java]: IIRC, Runnable was chosen back then because of the issues with using TWR which was generating warnings (because resource not accessed in body of TWR). I don't have strong opinions on this, other than we should only have one way to do things, rather than 2-3. Since the JDK has capabilities to acquire and release a session w/o a TWR and/or Runnable, one might also consider whether exposing acquire/release separately, and just use those.

File: src/java.base/share/classes/java/util/zip/Adler32.java
Comment (mcimadamore) [reply src/java.base/share/classes/java/util/zip/Adler32.java:105]: > > Although this looks much simpler and concise, it means "a new object is created for each invocation"
> 
> My comment was actually to see if DirectBuffer could extend AutoCloseable so that the acquire returns "this" for the non-session case. Doing the equivalent for the session case might leak into MemorySessionImpl implementing DirectBuffer which you probably don't want to do. If NOP_CLOSE.close can do the Reference.reachabilityFence(this) then it would at least improve some of the use-sites.

Not sure that is simpler. ByteBuffer <: AutoCloseable doesn't seem to make sense to me. I'm also not sure how much object allocation (all this stuff will become value types) should be the driving factor in these code paths.

Comment (AlanBateman) [reply src/java.base/share/classes/java/util/zip/Adler32.java:105]: Right, I didn't mean BB <: AC but to see if we avoid needing to wrap it because this PR is touching several low level and performance critical code paths. For the faraway places then having the close do a Reference.reachabilityFence(this) should avoid the surprise that the buffer has to kept alive even though it appears that the try-with-resources is doing it already.

File: src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java
Comment (mcimadamore) [inline src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java:914]: After looking at the implementation some more, I'm not sure this need fixing? E.g. this method is just using the address to compute some overlap - and return a buffer sliced accordingly. There's no access to the buffer data (except for the last part which does a `put`). The access will fail if the session is closed from underneath. I don't think this can crash the VM (in fact this code did not have a reachability fence to begin with).

File: src/java.base/share/classes/java/nio/Buffer.java
Comment (mcimadamore) [inline src/java.base/share/classes/java/nio/Buffer.java:827]: If allocation/performance is an issue, a relatively straightforward way to adjust the code would be to let go of the TWR entirely. E.g. we have code that does this:

```
Buffer b = ...
try {
    // use buffer.address();
} finally {
    Reference.reachabilityFence(b);
}
```

We could transform to this:

```
Buffer b = ...
acquire(b); // calls MemorySessionImpl.acquire0 if session is not null
try {
    // use buffer.address();
} finally {
    release(b); // calls MemorySessionImpl.release0 if session is not null (and maybe adds a reachability fence, just in case)
}
```
This leads to a simpler patch that is probably easier to validate. The remaining IOUtil code will be using a different idiom, and perhaps we can, at some point, go back and make that consistent too.

File: src/java.base/share/classes/sun/nio/fs/NativeBuffers.java
Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/fs/NativeBuffers.java]: > That is clear to me but I am trying to prevent future redundant guarding. Anyway, I will remove the comments.

The faraway use sites look much better now. The performance sensitive usages need close attention. Can you summarise the changes to DatagramChannel, are you creating a NoOpScopeAcquisition for every  send/receive? This is low level code where we do do something different to avoid the try-with-resources.

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (mcimadamore) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:590]: Why was the old code not using reachability fences? Bug or feature?

Comment (mcimadamore) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:590]: I see that there's a subsequent buffer call if `n > 0`, so that's probably why the fence was skipped? (I also assume that the code calling this method will access the buffer before/after, so reachability is never truly an issue - but for session-backed buffers this needs fixing).

Also, stepping back, I note how, if `receive0` was a native call using Linker, perhaps we wouldn't need all this manual address computation - we'd just get a memory segment slice from the buffer and pass that to the handle (which will perform the correct liveness check). E.g. maybe a better long term solution would be to panama-ize this code?

File: src/java.base/share/classes/sun/nio/ch/DirectBuffer.java
Comment (ExE-Boss) [inline src/java.base/share/classes/sun/nio/ch/DirectBuffer.java:43]: Again, updating this comment was missed:
```suggestion
    // try (var guard = NIO_ACCESS.acquireScope(bb)) {
    //     performOperation(guard.address());
    // }
```

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (AlanBateman) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:590]: Yes, once the memory/linker APIs are permanent then the SCTP implementation would be a good candidate to redo.

File: src/java.base/share/classes/java/nio/Buffer.java
Comment (AlanBateman) [reply src/java.base/share/classes/java/nio/Buffer.java:827]: The AutoCloseable experiment was interesting to try but I think you are right, acquire try { .. } finally release would be simpler and also remove the concerns about the performance critical code paths.

File: src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java
Comment (ExE-Boss) [reply src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java:914]: This `@SuppressWarnings` annotation is no longer needed:
```suggestion
```

File: src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java:125]: I think this looks better - but naming-wise it's still a bit problematic. This method really acquires the underlying session, not a scope. And also, here we have `OrNull`, but the already existing `acquireSession` also has a similar semantics w.r.t. null.

I suggest to rename:

* `acquireSession` -> `acquireSessionAsRunnable`
* `acquireScopeOrNull` -> `acquireSession`
* `releaseScope` -> `releaseSession`

Also, once we have `acquire/releaseSession`, it is not clear to me that we still need `acquireSessionAsRunnable` in the JavaNIOAccess class - it seems like you can create the Runnable where it's required (probably IOUtil), simply by using straight acquire/release.

Comment (minborg) [reply src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java:125]: The name "scope" was used in anticipation of the new proposed Java 20 naming. But I can change it back to session again. We could always rename later.

File: src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java
Comment (minborg) [reply src/java.base/share/classes/jdk/internal/access/JavaNioAccess.java:164]: Well spotted.

**Review by mcimadamore — APPROVED** [review]
Looks good - I'm glad that the changes to IOUtil turned out simpler than expected!

File: src/java.base/share/classes/sun/nio/ch/IOUtil.java
Comment (mcimadamore) [inline src/java.base/share/classes/sun/nio/ch/IOUtil.java:519]: Some (optional) style comments: I'm not sure this is a record. E.g. both this and the `Releaser` class are only used externally to call their `run` method. The fact that they are made up of components is immaterial to clients, which seems to suggest that an interface would be better - at least subjectively. If I were to write the code I will declare the implementation inside the factories, as local/anon classes.

Comment (mcimadamore) [reply src/java.base/share/classes/sun/nio/ch/IOUtil.java:519]: Separately, also (optional) stylistic: the indenting on this and following class is a bit odd. There is a certain tendency to compress lines - e.g. to reach for one-liners, when in reality the body is not that trivial. If I were to write the code I'd insert a newline after the `{` and I'd also break apart initialization (e.g. no two statements separated by `;` in the same line).

Also, I noted that you start the body of the record (e.g. `{`) on a new line, which I also find odd.

Comment (mcimadamore) [reply src/java.base/share/classes/sun/nio/ch/IOUtil.java:519]: (to be clear, some of the above comments refer to the code that was already there before your changes)

File: src/java.base/share/classes/sun/nio/ch/DirectBuffer.java
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/DirectBuffer.java:39]: Did you mean to leave the word "Guards" in the comment? I guess I would say something "See JavaNioAccess for methods to acquire/release" or something like that.

File: src/java.base/share/classes/sun/nio/ch/IOUtil.java
Comment (minborg) [reply src/java.base/share/classes/sun/nio/ch/IOUtil.java:519]: I also found the above as a bit odd but tried to stick with the existing implementation and style as much as possible.

File: src/java.base/share/classes/sun/nio/ch/IOUtil.java
Comment (minborg) [reply src/java.base/share/classes/sun/nio/ch/IOUtil.java:480]: This was the old behaviour which was retained in this PR.

Comment (ExE-Boss) [reply src/java.base/share/classes/sun/nio/ch/IOUtil.java:480]: I believe `UnsupportedOperationException` would be better for this.

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/IOUtil.java:480]: > I believe `UnsupportedOperationException` would be better for this.

It would but Per is right that this it is throwing ISE today, it's just that the PR adds an explicit check and that highlights that the wrong exception is thrown. Buffers that are views over memory that is thread confined can't be used with the AsynchronousXXXX APIs and this will needed to be specified. There is further work required in this area too because IOUtil was intended for synchronous I/O. The "async" support was added on a temporary basis and will need to be re-visited at some point.

**Comment by AlanBateman** [comment]
The latest version (30aff118) looks quite good but the naming at the use sites (`var scope = NIO_ACCESS.acquireSession(buffer);`) is temporarily confusing. Will that be fixed with the JEP  434 integrated or soon afterwards?

File: src/java.base/share/classes/java/util/zip/Deflater.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/Deflater.java:594]: Somewhat subjective but the original code, with inputAddress, was a easier to read and avoided having too much going on in the parameters to deflateBufferBytes. In any case, you probably should fix the formatting to avoid have different parameters with different indentations.

File: src/java.base/share/classes/java/util/zip/Deflater.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/Deflater.java:717]: Same here, I think I would keep outputAddress to keep the call to deflateByteBuffer simple.

File: src/java.base/share/classes/java/util/zip/Deflater.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/Deflater.java:740]: Another one where inputAddress and outputAddress have been removed so the call to deflatBufferBuffer is much hard to digest.

**Comment by minborg** [comment]
> The latest version ([30aff11](https://github.com/openjdk/jdk/commit/30aff1187d4978c59d1e6feebafc2d45aba3e5f2)) looks quite good but the naming at the use sites (`var scope = NIO_ACCESS.acquireSession(buffer);`) is temporarily confusing. Will that be fixed with the JEP 434 integrated or soon afterwards?

The name was selected in anticipation of the proposed renaming in JEP 434. I think once the JEP is merged, we can rename some of the methods (*Session* -> *Scope*) and then the user site naming would be much better. Another alternative, which certainly can be done, is to rename `scope` to `session` and then rename these back again when the JEP is merged. What is the preferred way in your opinion?

**Comment by AlanBateman** [comment]
> The name was selected in anticipation of the proposed renaming in JEP 434. I think once the JEP is merged, we can rename some of the methods (_Session_ -> _Scope_) and then the user site naming would be much better. Another alternative, which certainly can be done, is to rename `scope` to `session` and then rename these back again when the JEP is merged. What is the preferred way in your opinion?

Whatever is easiest. Given the fork for JDK 20 is coming up soon then it would be good to get the rename done soon after JEP 434 integrates, just to avoid JDK 20 having the mix of names.

**Review by AlanBateman — APPROVED** [review]
Thanks for persisting with it, latest version looks okay to me and the naming issue can be sorted out after the JEP is integrated.

File: src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java
Comment (mcimadamore) [reply src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java]: > Thanks for persisting with it, latest version looks okay to me and the naming issue can be sorted out after the JEP is integrated.


IMHO the renaming is not super important - the underlying abstraction managing the segment lifetime is still called MemorySession, even after https://git.openjdk.org/jdk/pull/10872. And, acquire/release are methods on MemorySession - so I think the current name might be more precise even after we integrate the latest API changes.

File: src/java.base/share/classes/java/nio/Buffer.java
Comment (PaulSandoz) [inline src/java.base/share/classes/java/nio/Buffer.java:838]: I prefer methods that do not expose the scope implementation so access is limited to just to the acquire/release methods, but i am unsure of the performance implications. These methods might not reliably inline, and we may need to ensure that first (which is also separately a good thing). I think it requires that the shared secret fields are stable and that there is only one implementation of `JavaNioAccess`, which there is, but we can enforce via sealing. Something to consider as a further iteration.

Comment (minborg) [reply src/java.base/share/classes/java/nio/Buffer.java:838]: > I prefer methods that do not expose the scope implementation so access is limited to just to the acquire/release methods, but i am unsure of the performance implications. These methods might not reliably inline, and we may need to ensure that first (which is also separately a good thing). I think it requires that the shared secret fields are stable and that there is only one implementation of `JavaNioAccess`, which there is, but we can enforce via sealing. Something to consider as a further iteration.

I agree. A similar approach was done in a previous incarnation of the PR whereby we exposed a `ScopeAcqusition` that held an internal copy of the buffer and provided means to release. We feared that would cause performance regression. But your comment made me think of another way of doing acquire/release which would not expose the session and that would always be performant. I will explore this idea a bit more.

**Review by mcimadamore — APPROVED** [review]
This looks more symmetric, I like it. Since `acquire` was only really returning null if the underlying buffer had no associated session, I believe this rewriting preserves the semantics we had before (e.g. no accidental double releases).

**Review by AlanBateman — APPROVED** [review]

**Review by PaulSandoz — APPROVED** [review]

**Review by bplb — APPROVED** [review]

**Comment by minborg** [comment]
I've converted this PR to a draft because we want to wait for the PR_20 branch being merged ([JEP ](https://openjdk.org/jeps/434)

**Pull request closed without merging**
